server/src/unifyfs_request_manager.c

```diff
@@ -2306,30 +2310,58 @@ static void chunk_read_response_rpc(hg_handle_t handle)
             hg_bulk_t bulk_handle;
             hret = margo_bulk_create(mid, 1, (void**)&resp_buf, &in.bulk_size,
                 HG_BULK_WRITE_ONLY, &bulk_handle);
-            assert(hret == HG_SUCCESS);
+            if (hret != HG_SUCCESS) {
+                LOGERR("failed to prepare bulk transfer");
+                ret = UNIFYFS_ERROR_MARGO;
+                goto out_respond;
+            }
 
             /* execute the transfer to pull data from remote side
-             * into our local bulk transfer buffer */
-            hret = margo_bulk_transfer(mid, HG_BULK_PULL, hgi->addr,
-                in.bulk_handle, 0, bulk_handle, 0, in.bulk_size);
-            assert(hret == HG_SUCCESS);
-
-            /* process read replies (headers and data) we just
-             * received */
-            rc = rm_post_chunk_read_responses(app_id, client_id,
-                src_rank, req_id, num_chks, bulk_sz, resp_buf);
-            if (rc != (int)UNIFYFS_SUCCESS) {
-                LOGERR("failed to handle chunk read responses")
-                ret = rc;
+             * into our local bulk transfer buffer.
+             * NOTE: mercury/margo bulk transfer does not check the maximum
+             * transfer size that the underlying transport supports, and a
+             * large bulk transfer may result in failure. */
+            int i = 0;
+            hg_size_t remain = in.bulk_size;
+
+            do {
+                hg_size_t offset = i * MAX_BULK_TX_SIZE;
+                hg_size_t len = remain < MAX_BULK_TX_SIZE
+                                ? remain : MAX_BULK_TX_SIZE;
+
+                hret = margo_bulk_transfer(mid, HG_BULK_PULL, hgi->addr,
+                                           in.bulk_handle, offset,
+                                           bulk_handle, offset, len);
+                if (hret != HG_SUCCESS) {
+                    break;
+                }
+
+                remain -= len;
+                i++;
+            } while (remain > 0);
+
+            if (hret == HG_SUCCESS) {
+                LOGDBG("transferred bulk data (%lu bytes)", in.bulk_size);
+
+                /* process read replies (headers and data) we just received */
+                rc = rm_post_chunk_read_responses(app_id, client_id,
+                        src_rank, req_id, num_chks, bulk_sz, resp_buf);
+                if (rc != (int)UNIFYFS_SUCCESS) {
+                    LOGERR("failed to handle chunk read responses")
+                        ret = rc;
+                }
+            } else {
+                LOGERR("failed to perform bulk transfer");
```

Comment:
Does it make sense to still post a read_response if the transfer fails, but indicate an error status value in the read request?  Would not posting anything lead to a hang while reading?

Reply:
@adammoody Yes, it would be better to inform client that the read has failed. I will update accordingly. Maybe, we will fail the whole operation as EIO, instead of returning partial data.

Reply:
@adammoody I've been looking at the code for handling read operations between the client and server, and I seems like we do not handle any server-side read failures properly at this moment (correct me if I am wrong). The following is the data structures that we have for handling reads:

```
typedef struct {
    size_t nbytes;      /* size of data chunk */
    size_t offset;      /* file offset */
    size_t log_offset;  /* remote log offset */
    int log_app_id;     /* remote log application id */
    int log_client_id;  /* remote log client id */
} chunk_read_req_t;

typedef struct {
    size_t offset;    /* file offset */
    size_t nbytes;    /* requested read size */
    ssize_t read_rc;  /* bytes read (or negative error code) */
} chunk_read_resp_t;

typedef struct {
    int rank;                /* remote delegator rank */
    int rdreq_id;            /* read-request id */
    int app_id;              /* app id of requesting client process */
    int client_id;           /* client id of requesting client process */
    int num_chunks;          /* number of chunk requests/responses */
    readreq_status_e status; /* summary status for chunk reads */
    size_t total_sz;         /* total size of data requested */
    chunk_read_req_t* reqs;  /* @RM: subarray of server_read_req_t.chunks
                              * @SM: received requests buffer */
    chunk_read_resp_t* resp; /* @RM: received responses buffer
                              * @SM: allocated responses buffer */
} remote_chunk_reads_t;

typedef struct {
    size_t length;  /* length of data to read */
    size_t offset;  /* file offset */
    int gfid;       /* global file id */
    int errcode;    /* request completion status */
} client_read_req_t;
```

I think we would need to set the error code to the `chunk_read_resp_t`, but the failure of the bulk transfer means that we have failed to receive this data structure from a remote delegator. I was initially thinking that setting the whole data fetch operations (`remote_chunk_reads_t`) from a specific remote server as a failure, but we would need to add an additional field (or defining a new enum in `readreq_status_e`, like READREQ_FAILED) to do so.

In any cases, the client-side code won't handle the error properly (`unifyfs_gfid_read_reqs`):

https://github.com/LLNL/UnifyFS/blob/cd0c9b45ea843b5af3e403526057488953bd092b/client/src/unifyfs.c#L1107

```

    /*
     * ToDo: Exception handling when some of the requests
     * are missed
     * */

```

So, my opinion is how about separating this problem (propagating server-side read I/O errors correctly to client) for another PR, because it will require some changes in server as well as client. In addition to the margo bulk transfer errors, there are other errors that we would need to handle properly in the `chunk_read_response_rpc` function, e.g., the code path for the `malloc(3)` failure will also skip reporting client, causing the client to wait until the timeout.

I can make a quick/dirty workaround to fail the client read, but also think that it would be better to be done more properly. Let me know how you think.